mono/metadata/object.c

```diff
@@ -1034,24 +1043,59 @@ mono_class_compute_gc_descriptor (MonoClass *klass)
 			gc_descr = mono_gc_make_descr_for_array (klass->byval_arg.type == MONO_TYPE_SZARRAY, bitmap, mono_array_element_size (klass) / sizeof (gpointer), mono_array_element_size (klass));
 			/*printf ("new vt array descriptor: 0x%x for %s.%s\n", class->gc_descr,
 				class->name_space, class->name);*/
-			if (bitmap != default_bitmap)
-				g_free (bitmap);
 		}
 	} else {
 		/*static int count = 0;
 		if (count++ > 58)
 			return;*/
 		bitmap = compute_class_bitmap (klass, default_bitmap, sizeof (default_bitmap) * 8, 0, &max_set, FALSE);
-		gc_descr = mono_gc_make_descr_for_object (bitmap, max_set + 1, klass->instance_size);
 		/*
 		if (class->gc_descr == MONO_GC_DESCRIPTOR_NULL)
 			g_print ("disabling typed alloc (%d) for %s.%s\n", max_set, class->name_space, class->name);
 		*/
 		/*printf ("new descriptor: %p 0x%x for %s.%s\n", class->gc_descr, bitmap [0], class->name_space, class->name);*/
-		if (bitmap != default_bitmap)
-			g_free (bitmap);
+
+		if (klass->has_weak_fields) {
+			gsize *weak_bitmap = NULL;
+			int weak_bitmap_nbits = 0;
+
+			weak_bitmap = (gsize *)mono_class_alloc0 (klass, klass->instance_size / sizeof (gsize));
+			if (mono_class_has_static_metadata (klass)) {
+				for (MonoClass *p = klass; p != NULL; p = p->parent) {
+					gpointer iter = NULL;
+					guint32 first_field_idx = mono_class_get_first_field_idx (p);
+					MonoClassField *field;
+
+					while ((field = mono_class_get_fields (p, &iter))) {
+						guint32 field_idx = first_field_idx + (field - p->fields);
+						if (MONO_TYPE_IS_REFERENCE (field->type) && mono_assembly_is_weak_field (p->image, field_idx + 1)) {
```

Comment:
We should also check for readonly flag

Reply:
why the readonly flag matters for this?

Reply:
It just feels wrong to me to have Weak on readonly field, the value should not be ever unset

Reply:
This PR doesn't include any form of error handling for bad field configurations. It's left for the C# compiler to eventually do it.

A readonly weak ref will work as advertised, only inited in the constructor and eventually zero'd when the object is gone. It's weird but kinda makes sense?

Reply:
I don't think C# work will be done anytime soon if ever. Maybe I misunderstand but `[Weak] readonly Foo f = Other.Value;` can be set back to null after initialization and Other.Value as an unique reference is gone which just looks odd to me.

Reply:
Hey Marek,

This PR is not the right medium to discuss semantics and scope. I support us discussing it all just not here.